lib/commons-core/src/main/java/org/apache/olingo/commons/core/edm/primitivetype/EdmTimeOfDay.java

```diff
@@ -19,103 +19,111 @@
-    } catch (final ClassCastException e) {
-      throw new EdmPrimitiveTypeException("The value type " + returnType + " is not supported.", e);
-    }
-  }
-
-  @Override
-  protected <T> String internalValueToString(final T value,
-      final Boolean isNullable, final Integer maxLength, final Integer precision,
-      final Integer scale, final Boolean isUnicode) throws EdmPrimitiveTypeException {
-
-    final Calendar dateTimeValue = EdmDateTimeOffset.createDateTime(value, true);
-
-    StringBuilder result = new StringBuilder();
-    EdmDateTimeOffset.appendTwoDigits(result, dateTimeValue.get(Calendar.HOUR_OF_DAY));
-    result.append(':');
-    EdmDateTimeOffset.appendTwoDigits(result, dateTimeValue.get(Calendar.MINUTE));
-    result.append(':');
-    EdmDateTimeOffset.appendTwoDigits(result, dateTimeValue.get(Calendar.SECOND));
-
-    final int fractionalSecs = value instanceof Timestamp ?
-        ((Timestamp) value).getNanos() :
-        dateTimeValue.get(Calendar.MILLISECOND);
-    try {
-      EdmDateTimeOffset.appendFractionalSeconds(result, fractionalSecs, value instanceof Timestamp, precision);
-    } catch (final IllegalArgumentException e) {
-      throw new EdmPrimitiveTypeException("The value '" + value + "' does not match the facets' constraints.", e);
-    }
-
-    return result.toString();
-  }
+	private static final LocalDate EPOCH = LocalDate.ofEpochDay(0l);
+	private static final EdmTimeOfDay INSTANCE = new EdmTimeOfDay();
+
+	public static EdmTimeOfDay getInstance() {
+		return INSTANCE;
+	}
+
+	@Override
+	public Class<?> getDefaultType() {
+		return Calendar.class;
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	protected <T> T internalValueOfString(final String value, final Boolean isNullable, final Integer maxLength,
+			final Integer precision, final Integer scale, final Boolean isUnicode, final Class<T> returnType)
+			throws EdmPrimitiveTypeException {
+		LocalTime time;
+		try {
+			time = LocalTime.parse(value);
+		} catch (DateTimeParseException ex) {
+			throw new EdmPrimitiveTypeException("The literal '" + value + "' has illegal content.");
+		}
+
+		// appropriate types
+		if (returnType.isAssignableFrom(LocalTime.class)) {
+			return (T) time;
+		} else if (returnType.isAssignableFrom(java.sql.Time.class)) {
+			return (T) java.sql.Time.valueOf(time);
+		}
+
+		// inappropriate types, which need to be supported for backward compatibility
+		ZonedDateTime zdt = LocalDateTime.of(EPOCH, time).atZone(ZoneId.systemDefault());
+		if (returnType.isAssignableFrom(Calendar.class)) {
+			return (T) GregorianCalendar.from(zdt);
+		} else if (returnType.isAssignableFrom(Long.class)) {
+			return (T) Long.valueOf(zdt.toInstant().toEpochMilli());
+		} else if (returnType.isAssignableFrom(Timestamp.class)) {
+			return (T) Timestamp.from(zdt.toInstant());
+		} else {
+			throw new EdmPrimitiveTypeException("The value type " + returnType + " is not supported.");
```

Comment:
we should handle return types java.util.Date and java.sql.Date

Reply:
java.util.Date we can add with the interpretation that we return a java.util.Date instance where the time components if interpreted in UTC represent the time of day.

But java.sql.Date _explicitly_ represents a date without time components but time of day is a time without date components. It really does not make sense to support java.sql.Date.